### PR TITLE
Integrate Rich into Accelerate

### PR DESCRIPTION
**Pull request by muellerzr**
This PR introduces handling the stack traces via [rich](https://github.com/Textualize/rich) project to help report a clean traceback if the user desires. It takes the very messy traceback and cleans up parts* of them

*more details below

## How to use:

Rich is a hard dep now in Accelerate and the right import is done in the base `__init__.py`, so this gets enabled by default!

## Example Tracebacks

Here are a few example tracebacks and what happened when.

### An exception from inside the script

In this example an exception was raised inside the `nlp_example.py` script where a `ValueError` was raised on a specific worker (hence why only one trace):

image

This is the entire trace, before only has what was reported to the logs. It's much cleaner and since it was an error inside here, the user does not need to know all the cuda bits

If the full trace is desired/needed, pass in `--debug` to `accelerate launch` to get the rest of it:

```bash
[20:42:05] WARNING  Sending process 361008 closing signal SIGTERM                                                                                                                              api.py:698
           ERROR    failed (exitcode: 1) local_rank: 1 (pid: 361009) of binary: /opt/conda/bin/python3.7                                                                                       api.py:672

Using --debug, `torch.distributed` Stack Trace:
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/zach_mueller_huggingface_co/accelerate/src/accelerate/commands/launch.py:460 in            │
│ multi_gpu_launcher                                                                               │
│                                                                                                  │
│ /opt/conda/lib/python3.7/site-packages/torch/distributed/run.py:755 in run                       │
│                                                                                                  │
│   752 │   elastic_launch(                                                                        │
│   753 │   │   config=config,                                                                     │
│   754 │   │   entrypoint=cmd,                                                                    │
│ ❱ 755 │   )(*cmd_args)                                                                           │
│   756                                                                                            │
│   757                                                                                            │
│   758 @record                                                                                    │
│                                                                                                  │
│ /opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py:131 in __call__         │
│                                                                                                  │
│   128 │   │   self._entrypoint = entrypoint                                                      │
│   129 │                                                                                          │
│   130 │   def __call__(self, *args):                                                             │
│ ❱ 131 │   │   return launch_agent(self._config, self._entrypoint, list(args))                    │
│   132                                                                                            │
│   133                                                                                            │
│   134 def _get_entrypoint_name(                                                                  │
│                                                                                                  │
│ /opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py:247 in launch_agent     │
│                                                                                                  │
│   244 │   │   │   # to the error file of the launcher process.                                   │
│   245 │   │   │   raise ChildFailedError(                                                        │
│   246 │   │   │   │   name=entrypoint_name,                                                      │
│ ❱ 247 │   │   │   │   failures=result.failures,                                                  │
│   248 │   │   │   )                                                                              │
│   249 │   │                                                                                      │
│   250 │   │   return result.return_values                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ChildFailedError: 
============================================================
examples/nlp_example.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2022-08-10_20:42:05
  host      : zach-multi-gpu.c.huggingface-ml.internal
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 361009)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

### A CUDA-specific exception

This example is the typical assert error that we get when using a label that doesn't line up with the right values (such as a label of `2` when doing binary classification)

This is where the * comes in, as the code traceback successfully shows what went wrong, but the CUDA report is still shown as well for debugging purposes. (Long traceback below with an image of the rich traceback smacked inbetween for sanity sake)

```bash
../aten/src/ATen/native/cuda/Loss.cu:271: nll_loss_forward_reduce_cuda_kernel_2d: block: [0,0,0], thread: [0,0,0] Assertion `t >= 0 && t < n_classes` failed.
../aten/src/ATen/native/cuda/Loss.cu:271: nll_loss_forward_reduce_cuda_kernel_2d: block: [0,0,0], thread: [0,0,0] Assertion `t >= 0 && t < n_classes` failed.
[W CUDAGuardImpl.h:113] Warning: CUDA warning: device-side assert triggered (function destroyEvent)
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Exception raised from create_event_internal at ../c10/cuda/CUDACachingAllocator.cpp:1387 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x3e (0x7f21ccfe61ee in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x23a21 (0x7f21f5686a21 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #2: c10::cuda::CUDACachingAllocator::raw_delete(void*) + 0x257 (0x7f21f568b977 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #3: <unknown function> + 0x478dd8 (0x7f221ebf5dd8 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #4: c10::TensorImpl::release_resources() + 0x175 (0x7f21ccfcd7a5 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #5: <unknown function> + 0x1735185 (0x7f21f6e86185 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #6: c10d::Reducer::~Reducer() + 0x1ef (0x7f21f9ea7aef in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #7: std::_Sp_counted_ptr<c10d::Reducer*, (__gnu_cxx::_Lock_policy)2>::_M_dispose() + 0x12 (0x7f221f1b24d2 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #8: <unknown function> + 0xa374b0 (0x7f221f1b44b0 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #9: <unknown function> + 0x37a13f (0x7f221eaf713f in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #10: <unknown function> + 0x37b05f (0x7f221eaf805f in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #11: <unknown function> + 0xdc31b (0x557671b6f31b in /opt/conda/bin/python3.7)
frame #12: <unknown function> + 0xf47e7 (0x557671b877e7 in /opt/conda/bin/python3.7)
frame #13: <unknown function> + 0x1d290e (0x557671c6590e in /opt/conda/bin/python3.7)
frame #14: PyGC_Collect + 0x49 (0x557671c65999 in /opt/conda/bin/python3.7)
frame #15: Py_FinalizeEx + 0x62 (0x557671cb81a2 in /opt/conda/bin/python3.7)
frame #16: <unknown function> + 0x24880b (0x557671cdb80b in /opt/conda/bin/python3.7)
frame #17: _Py_UnixMain + 0x3c (0x557671cdbb8c in /opt/conda/bin/python3.7)
frame #18: __libc_start_main + 0xea (0x7f223a1cbd0a in /lib/x86_64-linux-gnu/libc.so.6)
frame #19: <unknown function> + 0x1be51d (0x557671c5151d in /opt/conda/bin/python3.7)
```

image
image

```bash
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
[W CUDAGuardImpl.h:113] Warning: CUDA warning: device-side assert triggered (function destroyEvent)
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Exception raised from create_event_internal at ../c10/cuda/CUDACachingAllocator.cpp:1387 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x3e (0x7f4e94d151ee in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x23a21 (0x7f4ebd3b5a21 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #2: c10::cuda::CUDACachingAllocator::raw_delete(void*) + 0x257 (0x7f4ebd3ba977 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #3: <unknown function> + 0x478dd8 (0x7f4ee6924dd8 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #4: c10::TensorImpl::release_resources() + 0x175 (0x7f4e94cfc7a5 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #5: <unknown function> + 0x1735185 (0x7f4ebebb5185 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #6: c10d::Reducer::~Reducer() + 0x1ef (0x7f4ec1bd6aef in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #7: std::_Sp_counted_ptr<c10d::Reducer*, (__gnu_cxx::_Lock_policy)2>::_M_dispose() + 0x12 (0x7f4ee6ee14d2 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #8: <unknown function> + 0xa374b0 (0x7f4ee6ee34b0 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #9: <unknown function> + 0x37a13f (0x7f4ee682613f in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #10: <unknown function> + 0x37b05f (0x7f4ee682705f in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #11: <unknown function> + 0xdbe7c (0x5642f4b1ae7c in /opt/conda/bin/python3.7)
frame #12: <unknown function> + 0x179951 (0x5642f4bb8951 in /opt/conda/bin/python3.7)
frame #13: <unknown function> + 0xdbbff (0x5642f4b1abff in /opt/conda/bin/python3.7)
frame #14: <unknown function> + 0xda350 (0x5642f4b19350 in /opt/conda/bin/python3.7)
frame #15: <unknown function> + 0xdabb8 (0x5642f4b19bb8 in /opt/conda/bin/python3.7)
frame #16: <unknown function> + 0xdabce (0x5642f4b19bce in /opt/conda/bin/python3.7)
frame #17: _PyEval_EvalFrameDefault + 0x661a (0x5642f4bf23ca in /opt/conda/bin/python3.7)
frame #18: _PyFunction_FastCallKeywords + 0x184 (0x5642f4b771d4 in /opt/conda/bin/python3.7)
frame #19: <unknown function> + 0x13b0d8 (0x5642f4b7a0d8 in /opt/conda/bin/python3.7)
frame #20: _PyEval_EvalFrameDefault + 0xaab (0x5642f4bec85b in /opt/conda/bin/python3.7)
frame #21: _PyEval_EvalCodeWithName + 0x242 (0x5642f4b30ea2 in /opt/conda/bin/python3.7)
frame #22: PyEval_EvalCodeEx + 0x39 (0x5642f4b320b9 in /opt/conda/bin/python3.7)
frame #23: PyEval_EvalCode + 0x1b (0x5642f4c1115b in /opt/conda/bin/python3.7)
frame #24: <unknown function> + 0x23dd53 (0x5642f4c7cd53 in /opt/conda/bin/python3.7)
frame #25: PyRun_FileExFlags + 0x97 (0x5642f4c87307 in /opt/conda/bin/python3.7)
frame #26: PyRun_SimpleFileExFlags + 0x19c (0x5642f4c874dc in /opt/conda/bin/python3.7)
frame #27: <unknown function> + 0x248a39 (0x5642f4c87a39 in /opt/conda/bin/python3.7)
frame #28: _Py_UnixMain + 0x3c (0x5642f4c87b8c in /opt/conda/bin/python3.7)
frame #29: __libc_start_main + 0xea (0x7f4f01efad0a in /lib/x86_64-linux-gnu/libc.so.6)
frame #30: <unknown function> + 0x1be51d (0x5642f4bfd51d in /opt/conda/bin/python3.7)

ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: -6) local_rank: 0 (pid: 40380) of binary: /opt/conda/bin/python3.7
Traceback (most recent call last):
  File "/opt/conda/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 345, in wrapper
    return f(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/distributed/run.py", line 761, in main
    run(args)
  File "/opt/conda/lib/python3.7/site-packages/torch/distributed/run.py", line 755, in run
    )(*cmd_args)
  File "/opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py", line 131, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/conda/lib/python3.7/site-packages/torch/distributed/launcher/api.py", line 247, in launch_agent
    failures=result.failures,
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
======================================================
examples/nlp_example.py FAILED
------------------------------------------------------
Failures:
[1]:
  time      : 2022-08-08_19:58:40
  host      : zach-multi-gpu.c.huggingface-ml.internal
  rank      : 1 (local_rank: 1)
  exitcode  : -6 (pid: 40381)
  error_file: <N/A>
  traceback : Signal 6 (SIGABRT) received by PID 40381
------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2022-08-08_19:58:40
  host      : zach-multi-gpu.c.huggingface-ml.internal
  rank      : 0 (local_rank: 0)
  exitcode  : -6 (pid: 40380)
  error_file: <N/A>
  traceback : Signal 6 (SIGABRT) received by PID 40380
======================================================
Traceback (most recent call last):
  File "/opt/conda/bin/accelerate", line 33, in <module>
    sys.exit(load_entry_point('accelerate', 'console_scripts', 'accelerate')())
  File "/home/zach_mueller_huggingface_co/accelerate/src/accelerate/commands/accelerate_cli.py", line 43, in main
    args.func(args)
  File "/home/zach_mueller_huggingface_co/accelerate/src/accelerate/commands/launch.py", line 831, in launch_command
    multi_gpu_launcher(args)
  File "/home/zach_mueller_huggingface_co/accelerate/src/accelerate/commands/launch.py", line 450, in multi_gpu_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['torchrun', '--nproc_per_node', '2', 'examples/nlp_example.py']' returned non-zero exit status 1.
```
CUDA-specific errors like this I don't think we can clean up past this point, since that boils down into how `torch.distributed.launch` does things, but it helps clean up some bits!

## What this does not do:

This does *not* fix the duplicated code in the reported stack trace. The fix for this will come in a follow up PR, where I've drastically improved the logs for `torch.distributed` as well as how they get reported.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Looks nice, but as said offline, I think a `--debug` mode that executes the launch of everything under that context manager would probably be the most helpful.

**Comment by pacman100** [comment]
Nice utility! LGTM 🤗, I think Sylvain's suggestion would be a really good way to use this feature. Also, I think this won't result in rich traceback if the issue was in creating `Accelerate` object itself, right? if so, should that be mentioned as part of documentation?

**Review by pacman100 — APPROVED** [review]

File: examples/nlp_example.py [hunk 1]
```diff
@@ -97,6 +98,8 @@ def collate_fn(examples):
 def training_function(config, args):
     # Initialize accelerator
     accelerator = Accelerator(cpu=args.cpu, mixed_precision=args.mixed_precision)
+    if accelerator.process_index == 1:
+        raise ValueError()
```
Comment (sgugger) [inline examples/nlp_example.py:102]: I think this is a leftover of your tests?

File: setup.py [hunk 2]
```diff
@@ -50,7 +50,7 @@
         ]
     },
     python_requires=">=3.7.0",
-    install_requires=["numpy>=1.17", "packaging>=20.0", "psutil", "pyyaml", "torch>=1.4.0"],
+    install_requires=["numpy>=1.17", "packaging>=20.0", "psutil", "pyyaml", "torch>=1.4.0", "rich"],
```
Comment (sgugger) [inline setup.py:53]: You put `rich` in the core install here but use an `is_rich_available` later on.

File: src/accelerate/commands/launch.py [hunk 3]
```diff
@@ -444,10 +441,29 @@ def multi_gpu_launcher(args):
...
-    process = subprocess.Popen(cmd, env=current_env)
-    process.wait()
-    if process.returncode != 0:
-        raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
+    # use torchrun directly here and try to call it
+
+    with patch_environment(**current_env):
+        import torch.distributed.run as distrib_run
+        debug = getattr(args, "debug", False)
+        keys = """nnodes,nproc_per_node,rdzv_backend,rdzv_endpoint,rdzv_id,rdzv_conf,standalone,max_restarts,monitor_interval,start_method,role,module,m,no_python,run_path,log_dir,r,redirects,t,tee,node_rank,master_addr,master_port,training_script,training_script_args""".split(",")
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:449]: Should be its own list in a constant probably, so it's easier to read/add/see diff

**Review by sgugger — COMMENTED** [review]
Thanks for working more on this! There are a few things to clean up with the rich imports a bit all over the place. If it's a soft dependency they should be protected, if we want `rich` as a core dep it's different of course (but maybe let's roll a first version with it extra first?)

File: examples/nlp_example.py (hunk 1 above)
Comment (muellerzr) [reply examples/nlp_example.py]: Thanks, fixed this now! 🤗

File: setup.py (hunk 2 above)
Comment (muellerzr) [reply setup.py:53]: As discussed offline, opted to make rich a true dep in Accelerate

**Comment by pacman100** [comment]
In Multi GPU launcher, if one isn't using `--debug` then they won't see any error stack trace at all (might be hard to read but important for root cause and resolution). This makes `--debug` sort of mandatory instead of optional. Can the user have normal error traceback in `--debug` isn't given? Please let me know if I am missing something or way off.

**Pull request merged**